src/proxy-restore.h

```diff
@@ -3,21 +3,35 @@
 
 #include "vctrs-core.h"
 
-
-r_obj* vec_restore(r_obj* x, r_obj* to, enum vctrs_owned owned);
-r_obj* vec_restore_default(r_obj* x, r_obj* to, enum vctrs_owned owned);
-
-r_obj* vec_restore_recurse(r_obj* x, r_obj* to, enum vctrs_owned owned);
```

Comment:
I've changed the surface area of this API a bit:

- `vec_restore()` lost an `owned` arg and is only used for the "simple" non-recursive, non-owned case
- New `vec_restore_opts()` that we use most of the time
- `vec_restore_recurse()` is removed, use `vec_restore_opts()`